In [2]:
import threading 
from transformers import AutoTokenizer , AutoModelForCausalLM , TextIteratorStreamer
from time import time

C:\Users\DHIA\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\LocalCache\local-packages\Python312\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [3]:
tokenizer = AutoTokenizer.from_pretrained("gpt2")
model = AutoModelForCausalLM.from_pretrained("gpt2")

prompt = "Your Journey starts with one"
tokens = o=tokenizer.encode(prompt,return_tensors="pt")
tokens

tensor([[ 7120, 15120,  4940,   351,   530]])

In [4]:
def stream_output(use_cache):
    streamer = TextIteratorStreamer(tokenizer,skip_special_tokens=True)
    thread = threading.Thread(target=model.generate,kwargs={
        'input_ids':tokens,
        "max_new_tokens":100,
        "use_cache":use_cache,
        "streamer":streamer
    })
    
    thread.start()
    
    start_time = time()
    for token in streamer : 
        print(token,end="",flush=True)
    end_time = time()
    thread.join()
    
    elapsed = end_time - start_time
    print(f"\n\nUse cache = {use_cache} , Time taken : {elapsed:.3f} seconds\n")

In [5]:
print(f"\n\n### Without KEY-VALUE Caching ###")
stream_output(use_cache=False)

The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
The attention mask is not set and cannot be inferred from input because pad token is same as eos token. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.




### Without KEY-VALUE Caching ###
Your Journey starts with one of the most important things you can do to help your family.

You can help your family by:

Making sure you have a good job

Getting a job

Getting a job

Getting a job

Getting a job

Getting a job

Getting a job

Getting a job

Getting a job

Getting a job

Getting a job

Getting a job

Getting a job

Getting a job

Getting a

Use cache = False , Time taken : 10.845 seconds



In [6]:
print(f"\n\n### With KEY-VALUE Caching ###")
stream_output(use_cache=True)

The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.




### With KEY-VALUE Caching ###
Your Journey starts with one of the most important things you can do to help your family.

You can help your family by:

Making sure you have a good job

Getting a job

Getting a job

Getting a job

Getting a job

Getting a job

Getting a job

Getting a job

Getting a job

Getting a job

Getting a job

Getting a job

Getting a job

Getting a job

Getting a

Use cache = True , Time taken : 5.009 seconds



### Without KV Caching : 10.845 seconds
### With KV Caching    : 5.009 seconds